io/src/main/scala/fs2/io/tcp/SocketGroup.scala

```diff
@@ -0,0 +1,347 @@
+            asyncYield[F, Option[Long]] { cb =>
+              val start = System.currentTimeMillis()
+              ch.write(
+                buff,
+                remains,
+                TimeUnit.MILLISECONDS,
+                (),
+                new CompletionHandler[Integer, Unit] {
+                  def completed(result: Integer, attachment: Unit): Unit =
+                    cb(
+                      Right(
+                        if (buff.remaining() <= 0) None
+                        else Some(System.currentTimeMillis() - start)
+                      ))
+                  def failed(err: Throwable, attachment: Unit): Unit =
+                    cb(Left(err))
+                }
+              )
+            }.flatMap {
+              case None       => F.pure(())
+              case Some(took) => go(buff, (remains - took).max(0))
+            }
+
+          go(bytes.toBytes.toByteBuffer, timeout.map(_.toMillis).getOrElse(0l))
+        }
+
+        ///////////////////////////////////
+        ///////////////////////////////////
+
+        new Socket[F] {
+          def readN(numBytes: Int, timeout: Option[FiniteDuration]): F[Option[Chunk[Byte]]] =
+            readN0(numBytes, timeout)
+          def read(maxBytes: Int, timeout: Option[FiniteDuration]): F[Option[Chunk[Byte]]] =
+            read0(maxBytes, timeout)
+          def reads(maxBytes: Int, timeout: Option[FiniteDuration]): Stream[F, Byte] =
+            Stream.eval(read(maxBytes, timeout)).flatMap {
+              case Some(bytes) =>
+                Stream.chunk(bytes) ++ reads(maxBytes, timeout)
+              case None => Stream.empty
+            }
+
+          def write(bytes: Chunk[Byte], timeout: Option[FiniteDuration]): F[Unit] =
+            write0(bytes, timeout)
+          def writes(timeout: Option[FiniteDuration]): Pipe[F, Byte, Unit] =
+            _.chunks.flatMap { bs =>
+              Stream.eval(write(bs, timeout))
+            }
+
+          def localAddress: F[SocketAddress] =
+            blockingDelay(blockingExecutionContext)(ch.getLocalAddress)
+          def remoteAddress: F[SocketAddress] =
+            blockingDelay(blockingExecutionContext)(ch.getRemoteAddress)
+          def isOpen: F[Boolean] = blockingDelay(blockingExecutionContext)(ch.isOpen)
+          def close: F[Unit] = blockingDelay(blockingExecutionContext)(ch.close())
+          def endOfOutput: F[Unit] = blockingDelay(blockingExecutionContext) {
+            ch.shutdownOutput(); ()
+          }
+          def endOfInput: F[Unit] = blockingDelay(blockingExecutionContext) {
+            ch.shutdownInput(); ()
+          }
+        }
+      }
+    }
+    Resource.make(socket)(_ =>
+      blockingDelay(blockingExecutionContext)(if (ch.isOpen) ch.close else ()).attempt.void)
+  }
+}
+
+object SocketGroup {
+
+  def apply[F[_]: Sync: ContextShift](
+      blockingExecutionContext: ExecutionContext,
+      nonBlockingThreadCount: Int = 4,
```

Comment:
Why not 5? Should this be a function of available processors?

Reply:
Fixed, thanks